fix(qgsproject): add sip alias removeMapLayersById to removeMapLayers

benoitdm-oslandia opened the pull request:
Due to Qt5/Qt6 differences python test test_qgsproject::removeMapLayersByLayer did not compile (bad data type/function).

Funded By Oslandia

nyalldawson commented on python/PyQt6/core/auto_generated/project/qgsproject.sip.in:
```diff
@@ -1343,7 +1343,7 @@ should use :py:func:`~QgsProject.addMapLayers` instead.
 %End
 
-    void removeMapLayers( const QStringList &layerIds );
+    void removeMapLayers( const QStringList &layerIds ) /PyName=removeMapLayersById/;
```
Isn't this an API break? Maybe instead you'll need custom methodcode to check the list element types and then call the correct c++ method.

benoitdm-oslandia replied:
is that ok if I add a patching in `python/PyQt6/core/__init__.py.in`?

nyalldawson replied:
I'd personally prefer it was handled in method code, as that's compiled and we gain extra safety from it.

ptitjano replied:
I have reworked @benoitdm-oslandia's approach with a method code approach.

ptitjano committed:
qgsproject: Add a sip removeMapLayers  code to fix qt6 version

With this new method code, it is now possible to use
`QgsProject::removeMapLayers` with a list of layers or a list of
layers IDs in qt5 and qt6.

Co-authored-by: bdm-oslandia <benoit.de.mezzo@oslandia.com>

nyalldawson commented on src/core/project/qgsproject.h:
```diff
@@ -1404,6 +1405,57 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
+      sipIsErr = 1;
+      PyErr_SetString( PyExc_ValueError, "Expected a list of layers or layers IDs" );
+    }
+    else if ( PyList_Check( a0 ) && PyList_GET_SIZE( a0 ) )
```
```suggestion
    else if ( PyList_GET_SIZE( a0 ) )
```

nyalldawson commented on src/core/project/qgsproject.h:
```diff
@@ -1404,6 +1405,57 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
+        int state;
+        if ( sipCanConvertToType( firstLayerPyObj, sipType_QgsMapLayer, SIP_NOT_NONE ) )
+        {
+          const sipTypeDef *qlist_type = sipFindType( "QList<QgsVectorLayer *>" );
```
```suggestion
          const sipTypeDef *qlist_type = sipFindType( "QList<QgsMapLayer *>" );
```

Maybe add a test with non vector layers too?

ptitjano replied:
Done.

nyalldawson commented on src/core/project/qgsproject.h:
```diff
@@ -1404,6 +1405,57 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
+          if ( !sipIsErr )
+          {
+            sipCpp->removeMapLayers( *layersList );
+          }
```
What happens when an error occurs here? Is this covered by tests?

ptitjano replied:
A `TypeError` exception is raised. This is covered by tests.

nyalldawson commented on tests/src/python/test_qgsproject.py:
```diff
@@ -825,6 +822,19 @@ def test_removeMapLayerById(self):
+        l2_id = l2.id()
+        QgsProject.instance().removeMapLayers([l1_id, l2_id])
+        self.assertEqual(QgsProject.instance().count(), 0)
+
```
Add a test with mixed layers and strings, and non layer/string types

ptitjano replied:
Done.
